pkg/scaling/scale_handler.go

```diff
@@ -115,41 +118,140 @@ func (h *scaleHandler) DeleteScalableObject(scalableObject interface{}) error {
 }
 
 // startScaleLoop blocks forever and checks the scaledObject based on its pollingInterval
-func (h *scaleHandler) startScaleLoop(ctx context.Context, withTriggers *kedav1alpha1.WithTriggers, scalableObject interface{}) {
+func (h *scaleHandler) startScaleLoop(ctx context.Context, withTriggers *kedav1alpha1.WithTriggers, scalableObject interface{}, scalingMutex *sync.Mutex) {
 	logger := h.logger.WithValues("type", withTriggers.Kind, "namespace", withTriggers.Namespace, "name", withTriggers.Name)
 
 	// kick off one check to the scalers now
-	h.checkScalers(ctx, scalableObject)
+	h.checkScalers(ctx, scalableObject, scalingMutex)
 
 	pollingInterval := getPollingInterval(withTriggers)
 	logger.V(1).Info("Watching with pollingInterval", "PollingInterval", pollingInterval)
 
 	for {
 		select {
 		case <-time.After(pollingInterval):
-			h.checkScalers(ctx, scalableObject)
+			h.checkScalers(ctx, scalableObject, scalingMutex)
 		case <-ctx.Done():
 			logger.V(1).Info("Context canceled")
 			return
 		}
 	}
 }
 
+func (h *scaleHandler) startPushScalers(ctx context.Context, withTriggers *kedav1alpha1.WithTriggers, scalableObject interface{}, scalingMutex *sync.Mutex) {
+	logger := h.logger.WithValues("type", withTriggers.Kind, "namespace", withTriggers.Namespace, "name", withTriggers.Name)
+	ss, err := h.GetScalers(scalableObject)
+	if err != nil {
+		logger.Error(err, "Error getting scalers", "object", scalableObject)
+		return
+	}
+
+	for _, s := range ss {
+		scaler, ok := s.(scalers.PushScaler)
+		if !ok {
+			continue
+		}
+
+		go func() {
+			activeCh := make(chan bool)
+			go scaler.Run(ctx, activeCh)
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case active := <-activeCh:
+					scalingMutex.Lock()
+					switch obj := scalableObject.(type) {
+					case *kedav1alpha1.ScaledObject:
+						h.scaleExecutor.RequestScale(ctx, obj, active)
+					case *kedav1alpha1.ScaledJob:
+						// TODO: revisit when implementing ScaledJob
+						h.scaleExecutor.RequestJobScale(ctx, obj, active, 1, 1)
+					}
+					scalingMutex.Unlock()
+				}
+			}
+		}()
+	}
+}
+
 // checkScalers contains the main logic for the ScaleHandler scaling logic.
 // It'll check each trigger active status then call RequestScale
-func (h *scaleHandler) checkScalers(ctx context.Context, scalableObject interface{}) {
+func (h *scaleHandler) checkScalers(ctx context.Context, scalableObject interface{}, scalingMutex *sync.Mutex) {
 	scalers, err := h.GetScalers(scalableObject)
 	if err != nil {
 		h.logger.Error(err, "Error getting scalers", "object", scalableObject)
 		return
 	}
 
+	scalingMutex.Lock()
```

Comment:
I'd suspect that the vast majority of ScaledObjects will use pull based scalers, thus calling `Lock()` and using mutex every `checkScalers()` call seems to be unnecessary (expensive). We might want to change the logic a little bit and do the initial check for example in `HandleScalableObject()` to find out whether push scaler is used and accomodate the logic for it.
It's just an idea, I am thinking about a high scale scenario, where there are a lot of ScaledObjects to be handled by KEDA, so every optimization in this section could be benefitial. But maybe the resource cost benefits will not be enough to justify the increased code complexity. WDYT?

Reply:
I was thinking about this. There has to be some coordination between push scalers and pull scalers for the same scalableObject. The mutex here is per-scaled object mutex that's used to only synchronize calls between `startPushScalers` and `startPullScalers`. And in the majority case of pull scalers, `startPushScalers` will terminate, and this mutex should always be free and grabbing it is cheap. 

From `HandleScalableObject()` we can pass both start loops a channel that they push scale requests to. Is that what you were thinking of? I wasn't sure if one would perform better than the other. I tried to simulate it in this bench (assuming the general case of no lock contention)

```go
import (
	"sync"
	"testing"
)

func startUsingChannels(n int) {
	output := make(chan struct{})
	go func() {
		for i := 0; i < n; i++ {
			output <- struct{}{}
		}
		close(output)
	}()
	for range output { }
}

func startUsingMutex(n int) {
	mutex := sync.Mutex{}
	for i := 0; i < n; i++ {
		mutex.Lock()
		mutex.Unlock()
	}
}

func benchmarkFunc(i int, test func(int), b *testing.B) {
	for n := 0; n < b.N; n++ {
		test(i)
	}
}

func BenchmarkMutex1000(b *testing.B) { benchmarkFunc(1000, startUsingMutex, b) }
func BenchmarkMutex10000(b *testing.B) { benchmarkFunc(10000, startUsingMutex, b) }
func BenchmarkMutex100000(b *testing.B) { benchmarkFunc(100000, startUsingMutex, b) }

func BenchmarkChannel1000(b *testing.B) { benchmarkFunc(1000, startUsingChannels, b) }
func BenchmarkChannel10000(b *testing.B) { benchmarkFunc(10000, startUsingChannels, b) }
func BenchmarkChannel100000(b *testing.B) { benchmarkFunc(100000, startUsingChannels, b) }
```

```console
$ go test -bench=.
goos: linux
goarch: amd64
BenchmarkMutex1000-12             108013             10730 ns/op
BenchmarkMutex10000-12             10000            107060 ns/op
BenchmarkMutex100000-12             1104           1063382 ns/op
BenchmarkChannel1000-12             7610            161701 ns/op
BenchmarkChannel10000-12             763           1618556 ns/op
BenchmarkChannel100000-12             70          16365266 ns/op
```

I think go channels use mutexes to synchronize access to the channel. 

I'm not sure how to bench contention in a realistic way, but this naive change 

```go
func startUsingChannels(n int) {
	output := make(chan struct{})
	wg := sync.WaitGroup{}
	wg.Add(2)
	go func() {
		for i := 0; i < n; i++ {
			output <- struct{}{}
		}
		wg.Done()
	}()
	go func() {
		for i := 0; i < n; i++ {
			output <- struct{}{}
		}
		wg.Done()
	}()
	go func() {
		wg.Wait()
		close(output)
	}()
	for range output { }
}

func startUsingMutex(n int) {
	mutex := sync.Mutex{}
	go func() {
		for i := 0; i < n; i++ {
			mutex.Lock()
			mutex.Unlock()
		}
	}()
	go func() {
		for i := 0; i < n; i++ {
			mutex.Lock()
			mutex.Unlock()
		}
	}()
}
```

```console
$ go test -bench=.
goos: linux
goarch: amd64
BenchmarkMutex1000-12             113419             16936 ns/op
BenchmarkMutex10000-12             10000            140422 ns/op
BenchmarkMutex100000-12            10000            615928 ns/op
BenchmarkChannel1000-12             1090           1295513 ns/op
BenchmarkChannel10000-12             100          12669470 ns/op
BenchmarkChannel100000-12              9         123955100 ns/op
```

Reply:
Makes sense, thanks for the explanation!